File: src/ghsingo/internal/audio/drone.go
package audio

import (
	"math"

	"github.com/41490/chao5whistler/src/ghsingo/internal/composer"
)

// DroneVoice is a long-lived continuous oscillator: a sine at the active
// mode's root frequency layered with a sub octave, lightly detuned for
// movement. Amplitude is exponentially smoothed toward TargetGain so the
// drone reacts to the composer's slow state without clicks.
//
// The drone is the L0 floor of the ambient engine (#28/#31): it plays
// every sample, regardless of whether GH events are arriving.
type DroneVoice struct {
	sampleRate int
	phaseRoot  float64
	phaseSub   float64
	phaseLfo   float64
	rootHz     float64
	subHz      float64
	targetGain float32
	gain       float32
	gainSlew   float32
}

// NewDroneVoice initializes the drone in mode Yo at zero gain. Use
// SetMode + SetTargetGain to drive it from composer output.
func NewDroneVoice(sampleRate int) *DroneVoice {
	d := &DroneVoice{
		sampleRate: sampleRate,
		gainSlew:   1.0 / float32(sampleRate) * 4.0, // ~250 ms slew per unit
	}
	d.SetMode(composer.ModeYo)
	return d
}

// SetMode picks a root frequency based on the composer mode. Bright modes
// sit a touch higher; dark modes drop down.
func (d *DroneVoice) SetMode(m composer.Mode) {
	switch m {
	case composer.ModeYo:
		d.rootHz = 220.0 // A3
	case composer.ModeHira:
		d.rootHz = 196.0 // G3
	case composer.ModeIn:
		d.rootHz = 174.61 // F3
	case composer.ModeRyo:
		d.rootHz = 233.08 // Bb3
	default:
		d.rootHz = 220.0
	}
	d.subHz = d.rootHz * 0.5
}

// SetTargetGain sets the slew destination for the drone level.
func (d *DroneVoice) SetTargetGain(g float32) {
	if g < 0 {
		g = 0
	}
	if g > 1 {
		g = 1
	}
	d.targetGain = g
}

// NextSample advances one sample and returns the mono drone output.
func (d *DroneVoice) NextSample() float32 {
	if d.gain < d.targetGain {
		d.gain += d.gainSlew
		if d.gain > d.targetGain {
			d.gain = d.targetGain
		}
	} else if d.gain > d.targetGain {
		d.gain -= d.gainSlew
		if d.gain < d.targetGain {
			d.gain = d.targetGain
		}
	}

	dt := 1.0 / float64(d.sampleRate)
	d.phaseRoot += 2 * math.Pi * d.rootHz * dt
	d.phaseSub += 2 * math.Pi * d.subHz * dt
	d.phaseLfo += 2 * math.Pi * 0.1 * dt
	if d.phaseRoot > 2*math.Pi {
		d.phaseRoot -= 2 * math.Pi
	}
	if d.phaseSub > 2*math.Pi {
		d.phaseSub -= 2 * math.Pi
	}
	if d.phaseLfo > 2*math.Pi {
		d.phaseLfo -= 2 * math.Pi
	}

	lfo := 0.5 + 0.5*math.Sin(d.phaseLfo)
	root := math.Sin(d.phaseRoot) * 0.5
	sub := math.Sin(d.phaseSub) * 0.6
	mix := (root + sub) * (0.7 + 0.3*lfo)
	return float32(mix) * d.gain
}

File: src/rsghsing/src/audio/drone.rs
//! L0 continuous drone. Ported from `internal/audio/drone.go`.

use std::f64::consts::PI;

use crate::composer::Mode;

pub struct DroneVoice {
    sample_rate: i32,
    phase_root: f64,
    phase_sub: f64,
    phase_lfo: f64,
    root_hz: f64,
    sub_hz: f64,
    target_gain: f32,
    gain: f32,
    gain_slew: f32,
}

impl DroneVoice {
    pub fn new(sample_rate: i32) -> Self {
        let mut d = DroneVoice {
            sample_rate,
            phase_root: 0.0,
            phase_sub: 0.0,
            phase_lfo: 0.0,
            root_hz: 220.0,
            sub_hz: 110.0,
            target_gain: 0.0,
            gain: 0.0,
            gain_slew: 1.0 / sample_rate as f32 * 4.0, // ~250 ms slew per unit
        };
        d.set_mode(Mode::Yo);
        d
    }

    pub fn set_mode(&mut self, m: Mode) {
        self.root_hz = match m {
            Mode::Yo => 220.0,
            Mode::Hira => 196.0,
            Mode::In => 174.61,
            Mode::Ryo => 233.08,
        };
        self.sub_hz = self.root_hz * 0.5;
    }

    pub fn set_target_gain(&mut self, g: f32) {
        self.target_gain = g.clamp(0.0, 1.0);
    }

    pub fn next_sample(&mut self) -> f32 {
        if self.gain < self.target_gain {
            self.gain += self.gain_slew;
            if self.gain > self.target_gain {
                self.gain = self.target_gain;
            }
        } else if self.gain > self.target_gain {
            self.gain -= self.gain_slew;
            if self.gain < self.target_gain {
                self.gain = self.target_gain;
            }
        }

        let dt = 1.0 / f64::from(self.sample_rate);
        self.phase_root += 2.0 * PI * self.root_hz * dt;
        self.phase_sub += 2.0 * PI * self.sub_hz * dt;
        self.phase_lfo += 2.0 * PI * 0.1 * dt;
        if self.phase_root > 2.0 * PI {
            self.phase_root -= 2.0 * PI;
        }
        if self.phase_sub > 2.0 * PI {
            self.phase_sub -= 2.0 * PI;
        }
        if self.phase_lfo > 2.0 * PI {
            self.phase_lfo -= 2.0 * PI;
        }

        let lfo = 0.5 + 0.5 * self.phase_lfo.sin();
        let root = self.phase_root.sin() * 0.5;
        let sub = self.phase_sub.sin() * 0.6;
        let mix = (root + sub) * (0.7 + 0.3 * lfo);
        (mix as f32) * self.gain
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn gain_slews_toward_target_and_clamps() {
        let mut d = DroneVoice::new(44_100);
        d.set_target_gain(0.5);
        for _ in 0..44_100 {
            d.next_sample();
        }
        assert!((d.gain - 0.5).abs() < 1e-6);
        d.set_target_gain(9.0);
        assert_eq!(d.target_gain, 1.0);
    }

    #[test]
    fn mode_swaps_root_frequency() {
        let mut d = DroneVoice::new(44_100);
        d.set_mode(Mode::In);
        assert_eq!(d.root_hz, 174.61);
        assert_eq!(d.sub_hz, 174.61 * 0.5);
    }
}
